chain/consensus/mir/consensus.go

```diff
@@ -0,0 +1,191 @@
+	"github.com/filecoin-project/lotus/node/modules/dtypes"
+	xerrors "golang.org/x/xerrors"
+)
+
+var _ consensus.Consensus = &Mir{}
+
+var RewardFunc = func(ctx context.Context, vmi vm.Interface, em stmgr.ExecMonitor,
+	epoch abi.ChainEpoch, ts *types.TipSet, params []byte) error {
+	// TODO: No RewardFunc implemented for mir yet
+	return nil
+}
+
+type Mir struct {
+	store   *store.ChainStore
+	beacon  beacon.Schedule
+	sm      *stmgr.StateManager
+	genesis *types.TipSet
+}
+
+func NewConsensus(
+	ctx context.Context,
+	sm *stmgr.StateManager,
+	b beacon.Schedule,
+	g chain.Genesis,
+	netName dtypes.NetworkName,
+) (consensus.Consensus, error) {
+	return &Mir{
+		store:   sm.ChainStore(),
+		beacon:  b,
+		sm:      sm,
+		genesis: g,
+	}, nil
+}
+
+// CreateBlock creates a Filecoin block from the block template provided by Mir.
+func (bft *Mir) CreateBlock(ctx context.Context, w lapi.Wallet, bt *lapi.BlockTemplate) (*types.FullBlock, error) {
+	pts, err := bft.sm.ChainStore().LoadTipSet(ctx, bt.Parents)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load parent tipset: %w", err)
+	}
+
+	st, recpts, err := bft.sm.TipSetState(ctx, pts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load tipset state: %w", err)
+	}
+
+	next := &types.BlockHeader{
+		Miner:         builtin.SystemActorAddr, // Mir blocks are not signed, we use system addr as miner.
+		Parents:       bt.Parents.Cids(),
+		Ticket:        bt.Ticket,
+		ElectionProof: bt.Eproof,
+
+		BeaconEntries:         bt.BeaconValues,
+		Height:                bt.Epoch,
+		Timestamp:             bt.Timestamp,
+		WinPoStProof:          bt.WinningPoStProof,
+		ParentStateRoot:       st,
+		ParentMessageReceipts: recpts,
+	}
+	blsMessages, secpkMessages, err := consensus.MsgsFromBlockTemplate(ctx, bft.sm, next, pts, bt)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to process messages from block template: %w", err)
+	}
+
+	return &types.FullBlock{
+		Header:        next,
+		BlsMessages:   blsMessages,
+		SecpkMessages: secpkMessages,
+	}, nil
+}
+
+func (bft *Mir) ValidateBlockHeader(ctx context.Context, b *types.BlockHeader) (rejectReason string, err error) {
+	// TODO: This is missing.
+	log.Warn("oh oh! No specific block header validation implemented for Mir yet")
+	return "", nil
+}
+
+func (bft *Mir) ValidateBlock(ctx context.Context, b *types.FullBlock) (err error) {
+	log.Infof("starting block validation process at @%d", b.Header.Height)
+
+	if err := blockSanityChecks(b.Header); err != nil {
+		return xerrors.Errorf("incoming header failed basic sanity checks: %w", err)
+	}
+
+	h := b.Header
+
+	baseTs, err := bft.store.LoadTipSet(ctx, types.NewTipSetKey(h.Parents...))
+	if err != nil {
+		return xerrors.Errorf("load parent tipset failed (%s): %w", h.Parents, err)
+	}
+	if h.Height <= baseTs.Height() {
+		return xerrors.Errorf("block height not greater than parent height: %d != %d", h.Height, baseTs.Height())
+	}
+
+	// Allow a small block drift
```

Comment:
I don't think this is critical is critical. Blocks currently include the timestamps when the Lotus block was assembled to prevent validators from pushing blocks in the future or too in the past. I'd say that that's enough for now.

Reply:
And is the block hash computed over its timestamp as well? Because if that's the case, if n validators attach n different timestamps to the same block they receive from Mir, Lotus might see that as n different blocks. Is that the case?

Reply:
Good catch. As discussed sync, until we have a way to assign a single timestamp to the batch exported by Mir so all validators end up with the same block, we will add the height to the timestamp field of the block.

I will add an issue to track this.

Reply:
https://github.com/consensus-shipyard/lotus/issues/15